### PR TITLE
Add templated probing path to store for developer scenarios

### PR DESCRIPTION
**Pull request by ramarag**
This is a followup of dotnet/core-setup#2539
Fixes https://github.com/dotnet/sdk/issues/1298

@eerhardt @gkhanna79 PTAL

### CODE REVIEW

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.ComposeStore.targets
Comment (gkhanna79) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.ComposeStore.targets]: Where does Store output folder get the architecture now from?

Comment (gkhanna79) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.ComposeStore.targets]: Why are we removing this? I would not have expected this to happen.

Comment (ramarag) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.ComposeStore.targets]: Previously userprofile store was in
<user_profile>\<arch>\store

since we have removed both sdk and shared Fx from user_profile, this location no longer makes sense. I have moved store now under <user_profile>

The tooling adds <Arch> and <tfm> and passes this path as an additional probing path for developer scenarios

**Comment by gkhanna79** [comment]
I see. the arch under store subfolder remains, right?




Thanks,
Gaurav


-------- Original message --------

**Comment by ramarag** [comment]
@dotnet-bot test Windows_NT_FullFramework Release

**Comment by ramarag** [comment]
@dotnet-bot test Windows_NT Release
@dotnet-bot  test Windows_NT_FullFramework Release

**Comment by ramarag** [comment]
cc @livarcocc @nguerrera

**Comment by gkhanna79** [comment]
Change lgtm. Eric and Livar, can you ptal?




Thanks,
Gaurav


-------- Original message --------

**Review by gkhanna79 — APPROVED** [review]
@eerhardt PTAL as well.

File: test/Microsoft.NET.Build.Tests/GivenThatWeWantToBuildANetCoreApp.cs
Comment (eerhardt) [inline test/Microsoft.NET.Build.Tests/GivenThatWeWantToBuildANetCoreApp.cs]: This should assert that the file exists, not silently ignore when the file doesn't exist.

File: src/Tasks/Microsoft.NET.Build.Tasks/GenerateRuntimeConfigurationFiles.cs
Comment (eerhardt) [inline src/Tasks/Microsoft.NET.Build.Tasks/GenerateRuntimeConfigurationFiles.cs]: This feels odd that we are passing part of the path in on a parameter named `DeveloperStoreLocation`, and then the C# code appends more folders to the path.

Instead, I would expect the whole path to be built in one location (the MSBuild targets), and then passed in.  That way if, for whatever reason, someone can modify the whole path, if necessary.

Also, maybe it would be more extensible to have the parameter be "AdditionalProbingPaths" and take a list of paths.  Then the MSBuild targets just passes in one, but we can add more later (or someone else can).  Thoughts?

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.targets
Comment (eerhardt) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.targets]: Does it make sense to reuse `DefaultComposeDir` here?  Instead of duplicating this location between here and ComposeStore.targets.

**Comment by gkhanna79** [comment]
@ramarag Can you get the feedback incorporated so that we can get the PR ready for approval?

**Comment by eerhardt** [comment]
If necessary, I can make the updates if @ramarag doesn't have the time.

**Comment by gkhanna79** [comment]
@ramarag Will you be able to get the updates done by 1pm today? If not, @eerhardt it will be great if you can get them going.

**Comment by gkhanna79** [comment]
@eerhardt Will you be able to get the changes for this?

**Comment by eerhardt** [comment]
Just started working on them now.

**Comment by bleroy** [comment]
Does this need to be reflected in documentation?

**Comment by eerhardt** [comment]
> Does this need to be reflected in documentation?

Do we have the multi-level lookup locations listed in the documentation?  If so, yes.  By default, the host only knows how to look next to itself and in the "global" (ProgramFiles) location for 'runtime stores'.

We removed the `%USERPROFILE%` location from the host and only add that location when you `dotnet run` or `F5` a project.  That location is not used in published applications.

**Comment by bleroy** [comment]
> Do we have the multi-level lookup locations listed in the documentation?

Currently, we don't, so I'll leave it that way, at least for now. Thanks.

**Comment by gkhanna79** [comment]
> Currently, we don't, so I'll leave it that way, at least for now.

See Proposed Changes section at https://github.com/dotnet/core-setup/blob/master/Documentation/design-docs/multilevel-sharedfx-lookup.md - they are documented.

**Comment by bleroy** [comment]
Thanks, this is great. I meant on the docs.microsoft.com topic.

File: src/Tasks/Microsoft.NET.Build.Tasks/GenerateRuntimeConfigurationFiles.cs
Comment (ramarag) [inline src/Tasks/Microsoft.NET.Build.Tasks/GenerateRuntimeConfigurationFiles.cs]: I think you can use string[] here instead

**Comment by ramarag** [comment]
LGTM

**Review by nguerrera — APPROVED** [review]

**Comment by gkhanna79** [comment]
@eerhardt @nguerrera @livarcocc What are the next steps to get this merged?

File: src/Tasks/Microsoft.NET.Build.Tasks/GenerateRuntimeConfigurationFiles.cs
Comment (eerhardt) [inline src/Tasks/Microsoft.NET.Build.Tasks/GenerateRuntimeConfigurationFiles.cs]: I have a slight preference to keeping ITaskItem in case we ever want to add metadata to these items.

**Comment by eerhardt** [comment]
@MattGertz for approval

**Customer scenario**

With the introduction of the `dotnet store` feature, we first allowed the host to look in `%USERPROFILE%\.dotnet` by default all the time.  This way a developer could run `dotnet store` for a few dependencies, and then `dotnet run` their app, and the crossgen'd dependencies would be picked up from `%USERPROFILE%`.
We have since removed that location from the host because in production scenarios, we don't want it probing `%USERPROFILE%`.
This change adds the `%USERPROFILE%` location to be probed only at "dev-time" scenarios like `dotnet run` by adding it to the runtimeconfig.dev.json file, which isn't used in published apps.

**Bugs this fixes:** 

https://github.com/dotnet/sdk/issues/1298

**Workarounds, if any**

Developers would have to manually pass the `%USERPROFILE%\.dotnet\store` location on the command line every time they wanted to use a runtime store created by `dotnet store`.

**Risk**

Low - this re-enables the behavior that the host used to have, but only in dev time scenarios.

**Performance impact**

One new location is probed during `dotnet run`.
Developers can use `dotnet store` to crossgen their dependencies and get their projects executed faster using `dotnet run`.

**Is this a regression from a previous update?**

This behavior was consciously removed from the host and added to the SDK.

**Root cause analysis:**

This behavior was flagged as part of a security push for the new multi-level lookup feature.

**How was the bug found?**

Design/security review.

**Comment by MattGertz** [comment]
I agree & will take to JoC today.

**Comment by livarcocc** [comment]
@MattGertz for approval. This is not a in-box change Matt.

**Pull request merged**
